dvc/fs/dvc.py

```diff
@@ -121,11 +123,13 @@ def __init__(
         from pygtrie import Trie
 
         super().__init__()
+        self._repo_stack = ExitStack()
         if repo is None:
             repo = self._make_repo(url=url, rev=rev, subrepos=subrepos, **repo_kwargs)
             assert repo is not None
             # pylint: disable=protected-access
             repo_factory = repo._fs_conf["repo_factory"]
+            self._repo_stack.enter_context(repo)
```

Comment:
Any time that dvcfs constructs a new repo (or subrepo) instance we need to ensure that it gets closed cleanly, otherwise the index context for that subrepo could end up being opened/locked and never closed/unlocked